source/common/http/conn_manager_utility.cc

```diff
@@ -36,12 +36,17 @@ void ConnectionManagerUtility::mutateRequestHeaders(Http::HeaderMap& request_hea
                                                     Runtime::Loader& runtime,
                                                     const LocalInfo::LocalInfo& local_info) {
   // Clean proxy headers.
-  request_headers.removeConnection();
   request_headers.removeEnvoyInternalRequest();
   request_headers.removeKeepAlive();
   request_headers.removeProxyConnection();
   request_headers.removeTransferEncoding();
-  request_headers.removeUpgrade();
+
+  // If this is a WebSocket Upgrade request, do not remove the Connection and Upgrade headers,
+  // as we forward them verbatim to the upstream hosts.
+  if (!Utility::isWebSocketUpgradeRequest(request_headers)) {
```

Comment:
Good point. Let me look around for something that would tell me whether the request is HTTP1 or 2 and do this check.

On the performance issue, in the full websocket PR, the header lookup is done once again in the connection manager to kick off the websocket related code. I had thought about avoiding the double lookup for HTTP1, but it would require passing a "is_websocket" boolean parameter to this function. If you guys are okay with chaning the function signature and putting up with some clumsiness, then we can avoid the double header lookup for HTTP1.

See here https://github.com/lyft/envoy/compare/master...amalgam8:envoy-websocket#diff-d70ea2d9706e0246700148b2e2bb63ceR501 for reference.